runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -32,13 +57,31 @@ struct iree_hal_cuda_tracing_context_t {
   // between events and we need a stable base event.
   CUevent base_event;
 
-  // Indices into |event_pool| defining a ringbuffer.
-  uint32_t query_head;
-  uint32_t query_tail;
+  // Unallocated events.
+  iree_hal_cuda_tracing_context_event_t* event_freelist_head;
+
+  // Submitted events.
+  iree_hal_cuda_tracing_context_event_list_t submitted_event_list;
+
   uint32_t query_capacity;
 
   // Event pool reused to capture tracing timestamps.
-  CUevent event_pool[IREE_HAL_CUDA_TRACING_DEFAULT_QUERY_CAPACITY];
+  // The lifetime of the Events are as follows.
+  // 1) All events are allocated when the tracing context is created.
+  // 2) When a command_buffer inserts a query via:
+  //     iree_hal_cuda_**_tracing_context_insert_query
```

Comment:
Style nit: we don't ident further after the first line of an bullet--just need to keep starting at the same column.